only clean up iptables chains periodically in large clusters

danwinship opened the pull request:
#### What type of PR is this?
/kind cleanup
/kind feature

#### What this PR does / why we need it:
Followup to #110328 (which this branch includes). This makes it so that in "large" clusters, we only clean up stale service chains once a minute, rather than doing it on every sync. In very large clusters, this cuts several seconds off each `syncProxyRules` run.

Since the chains being removed by this code are not referenced by any other rules, it doesn't matter if we let them sit around for a while before we get around to deleting them; they're not going to have any effect on packet processing either way.

However, since it could be confusing if an admin does `iptables-save` and finds rules that are no longer in use (eg, a `KUBE-SEP-` chain for a pod that exited), I made it so that it only switches from "synchronous deletion" to "periodic deletion" in "large" clusters, using the same definition of "large" as we use for deciding when to start eliminating comments from the iptables rules.

#### Which issue(s) this PR fixes:
none

#### Does this PR introduce a user-facing change?
```release-note
In "large" clusters, kube-proxy in iptables mode will now sometimes
leave unused rules in iptables for a while (up to `--iptables-sync-period`)
before deleting them. This improves performance by not requiring it to
check for stale rules on every sync. (In smaller clusters, it will still
remove unused rules immediately once they are no longer used.)

(The threshold for "large" used here is currently "1000 endpoints" but
this is subject to change.)
```

/sig network
/priority important-longterm

danwinship committed:
proxy/iptables: only clean up chains periodically in large clusters

"iptables-save" takes several seconds to run on machines with lots of
iptables rules, and we only use its result to figure out which chains
are no longer referenced by any rules. While it makes things less
confusing if we delete unused chains immediately, it's not actually
_necessary_ since they never get called during packet processing. So
in large clusters, make it so we only clean up chains periodically
rather than on every sync.

danwinship committed:
proxy/iptables: abstract the "endpointChainsNumberThreshold" a bit

Turn this into a generic "large cluster mode" that determines whether
we optimize for performance or debuggability.

aojea commented:
just to try to understand these series of PR, what is the scale problem we are addressing?
memory consumption? time to sync penalty of running iptables-save?

danwinship commented:
Speed. `iptables-save` gets really slow when you have very very many iptables rules.

danwinship committed:
proxy/iptables: Add a unit test with multiple resyncs

aojea commented on pkg/util/iptables/save_restore.go:
```diff
@@ -21,102 +21,32 @@ import (
+// GetChains parses iptables-save data to find the chains that are defined.
+// It assumes that save contains a single table's data, and returns a map with
+// keys for every chain defined in that table.
+func GetChains(save []byte) map[Chain]struct{} {
```
`GetChainsFromTable` ??

aojea commented on pkg/util/iptables/save_restore.go:
```diff
@@ -21,102 +21,32 @@ import (
+func GetChains(save []byte) map[Chain]struct{} {
+	chainsMap := make(map[Chain]struct{})
+
+	for {
```
this is actually much simpler, 
```
find \n:
get the part between that and the blank space
repeat
```

aojea commented on pkg/proxy/iptables/proxier.go:
```diff
@@ -404,6 +405,8 @@ var iptablesCleanupOnlyChains = []iptablesJumpChain{
 	{utiliptables.TableFilter, kubeServicesChain, utiliptables.ChainInput, "kubernetes service portals", []string{"-m", "conntrack", "--ctstate", "NEW"}},
 }
 
+var iptablesCleanupPeriod = time.Minute
```
what if we only do the cleanup during the `syncPeriod` ?

```
iptables:
  masqueradeAll: false
  masqueradeBit: 14
  minSyncPeriod: 1s
  syncPeriod: 30s
```

this way it can be parameterizable

danwinship replied:
makes sense

aojea commented on pkg/proxy/iptables/proxier.go:
```diff
@@ -1339,19 +1331,34 @@ func (proxier *Proxier) syncProxyRules() {
+					// We must (as per iptables) write a chain-line
+					// for it, which has the nice effect of flushing
+					// the chain. Then we can remove the chain.
+					proxier.natChains.Write(utiliptables.MakeChainLine(chain))
+					proxier.natRules.Write("-X", chainString)
```
we should be able to add an unit test, 
```diff
diff --git a/pkg/proxy/iptables/proxier_test.go b/pkg/proxy/iptables/proxier_test.go
index 100d5cfa609..961720cc7d7 100644
--- a/pkg/proxy/iptables/proxier_test.go
+++ b/pkg/proxy/iptables/proxier_test.go
@@ -7698,3 +7698,104 @@ func TestNoEndpointsMetric(t *testing.T) {
                })
        }
 }
+
+func TestLargeClustersIPtables(t *testing.T) {
+       ipt := iptablestest.NewFake()
+       fp := NewFakeProxier(ipt)
+
+       tcpProtocol := v1.ProtocolTCP
+       makeServiceMap(fp,
+               // create ClusterIP service
+               makeTestService("ns1", "svc1", func(svc *v1.Service) {
+                       svc.Spec.ClusterIP = "172.30.0.41"
+                       svc.Spec.Ports = []v1.ServicePort{{
+                               Name:     "p80",
+                               Port:     80,
+                               Protocol: v1.ProtocolTCP,
+                       }}
+               }),
+       )
+
+       eps := makeTestEndpointSlice("ns1", "svc1", 1, func(eps *discovery.EndpointSlice) {
+               eps.AddressType = discovery.AddressTypeIPv4
+               eps.Endpoints = []discovery.Endpoint{{
+                       Addresses: []string{"10.180.0.1"},
+               }}
+               eps.Ports = []discovery.EndpointPort{{
+                       Name:     utilpointer.StringPtr("p80"),
+                       Port:     utilpointer.Int32(80),
+                       Protocol: &tcpProtocol,
+               }}
+       })
+       populateEndpointSlices(fp, eps)
+
+       fp.lastIPTablesCleanup = time.Now()
+       fp.syncProxyRules()
+
+       expected := dedent.Dedent(`
+               *filter
+               :KUBE-NODEPORTS - [0:0]
+               :KUBE-SERVICES - [0:0]
+               :KUBE-EXTERNAL-SERVICES - [0:0]
+               :KUBE-FORWARD - [0:0]
+               -A KUBE-FORWARD -m conntrack --ctstate INVALID -j DROP
+               -A KUBE-FORWARD -m comment --comment "kubernetes forwarding rules" -m mark --mark 0x4000/0x4000 -j ACCEPT
+               -A KUBE-FORWARD -m comment --comment "kubernetes forwarding conntrack rule" -m conntrack --ctstate RELATED,ESTABLISHED -j ACCEPT
+               COMMIT
+               *nat
+               :KUBE-NODEPORTS - [0:0]
+               :KUBE-SERVICES - [0:0]
+               :KUBE-MARK-MASQ - [0:0]
+               :KUBE-POSTROUTING - [0:0]
+               :KUBE-SEP-SXIVWICOYRO3J4NJ - [0:0]
+               :KUBE-SVC-XPGD46QRK7WJZT7O - [0:0]
+               -A KUBE-SERVICES -m comment --comment "ns1/svc1:p80 cluster IP" -m tcp -p tcp -d 172.30.0.41 --dport 80 -j KUBE-SVC-XPGD46QRK7WJZT7O
+               -A KUBE-SERVICES -m comment --comment "kubernetes service nodeports; NOTE: this must be the last rule in this chain" -m addrtype --dst-type LOCAL -j KUBE-NODEPORTS
+               -A KUBE-MARK-MASQ -j MARK --or-mark 0x4000
+               -A KUBE-POSTROUTING -m mark ! --mark 0x4000/0x4000 -j RETURN
+               -A KUBE-POSTROUTING -j MARK --xor-mark 0x4000
+               -A KUBE-POSTROUTING -m comment --comment "kubernetes service traffic requiring SNAT" -j MASQUERADE
+               -A KUBE-SEP-SXIVWICOYRO3J4NJ -m comment --comment ns1/svc1:p80 -s 10.180.0.1 -j KUBE-MARK-MASQ
+               -A KUBE-SEP-SXIVWICOYRO3J4NJ -m comment --comment ns1/svc1:p80 -m tcp -p tcp -j DNAT --to-destination 10.180.0.1:80
+               -A KUBE-SVC-XPGD46QRK7WJZT7O -m comment --comment "ns1/svc1:p80 cluster IP" -m tcp -p tcp -d 172.30.0.41 --dport 80 ! -s 10.0.0.0/8 -j KUBE-MARK-MASQ
+               -A KUBE-SVC-XPGD46QRK7WJZT7O -m comment --comment "ns1/svc1:p80 -> 10.180.0.1:80" -j KUBE-SEP-SXIVWICOYRO3J4NJ
+               COMMIT
+               `)
+
+       assertIPTablesRulesEqual(t, getLine(), expected, fp.iptablesData.String())
+
+       fp.lastIPTablesCleanup = time.Now()
+       fp.largeClusterMode = true
+       fp.OnEndpointSliceDelete(eps)
+       fmt.Println("sync 2")
+       fp.syncProxyRules()
+
+       expected = dedent.Dedent(`
+               *filter
+               :KUBE-NODEPORTS - [0:0]
+               :KUBE-SERVICES - [0:0]
+               :KUBE-EXTERNAL-SERVICES - [0:0]
+               :KUBE-FORWARD - [0:0]
+               -A KUBE-SERVICES -m comment --comment "ns1/svc1:p80 has no endpoints" -m tcp -p tcp -d 172.30.0.41 --dport 80 -j REJECT
+               -A KUBE-FORWARD -m conntrack --ctstate INVALID -j DROP
+               -A KUBE-FORWARD -m comment --comment "kubernetes forwarding rules" -m mark --mark 0x4000/0x4000 -j ACCEPT
+               -A KUBE-FORWARD -m comment --comment "kubernetes forwarding conntrack rule" -m conntrack --ctstate RELATED,ESTABLISHED -j ACCEPT
+               COMMIT
+               *nat
+               :KUBE-NODEPORTS - [0:0]
+               :KUBE-SERVICES - [0:0]
+               :KUBE-MARK-MASQ - [0:0]
+               :KUBE-POSTROUTING - [0:0]
+               :KUBE-SEP-SXIVWICOYRO3J4NJ - [0:0]
+               :KUBE-SVC-XPGD46QRK7WJZT7O - [0:0]
+               -A KUBE-SERVICES -m comment --comment "kubernetes service nodeports; NOTE: this must be the last rule in this chain" -m addrtype --dst-type LOCAL -j KUBE-NODEPORTS
+               -A KUBE-MARK-MASQ -j MARK --or-mark 0x4000
+               -A KUBE-POSTROUTING -m mark ! --mark 0x4000/0x4000 -j RETURN
+               -A KUBE-POSTROUTING -j MARK --xor-mark 0x4000
+               -A KUBE-POSTROUTING -m comment --comment "kubernetes service traffic requiring SNAT" -j MASQUERADE
+               COMMIT
+               `)
+
+       assertIPTablesRulesEqual(t, getLine(), expected, fp.iptablesData.String())
+
+}
```

aojea replied:
it should check that adds the `-X` rules or not

aojea commented:
+1
Just one consideration, to tie the behavior to the `syncPeriod` option instead of adding a new period
Also, I think we should have a test to check that -X rules are added or removed

/assign @thockin

danwinship commented:
ok, rebased for `iptables-counters` fixes, changed to use `syncPeriod` as the timeout, and added unit tests.

I pulled in https://github.com/kubernetes/kubernetes/pull/110334/commits/c12da17838a91917d379f4cc52079c417241530a from #110268, which serves the purpose here of confirming that chain deletion behavior is unchanged in "small" clusters. Then I renamed `TestEndpointCommentElision` to `TestSyncProxyRulesLargeClusterMode` and extended it to test the behavior of service deletion when `largeClusterMode` is triggered to confirm that as well.

aojea commented:
> k8s.io/kubernetes/pkg/controller/volume/attachdetach/reconciler: Test_Run_OneVolumeDetachFailNodeWithReadWriteOnce expand_less	

/test pull-kubernetes-unit
